File: 그래프/BOJ-1647.py
# BOJ-1647.py
# 도시 분할 계획

import heapq

def prim(node):
    heap = [] # 빠르게 최솟값을 찾기 위한 heap (리스트형태)
    visited[node] = True
    total = 0
    count = 1
    max_cost = 0

    # 해당 노드에 연결된 연결선을 추가한다.
    for a in adj[node]:
        heapq.heappush(heap, a)
       
    while heap:
        cost, node = heapq.heappop(heap) # heap 중 최소 비용의 연결을 반환
        if not visited[node]:
            print(cost, node)
            visited[node] = True
            max_cost = max(max_cost, cost)
            total += cost
            count += 1

            # 해당 노드에 연결된 연결선을 추가한다.
            for a in adj[node]:
                heapq.heappush(heap, a)
        
        # 연결선이 N개 선택된 경우 종료한다.
        if count == N:
            return total - max_cost
        
N, M = map(int, input().split())

# 인접 노드를 저장하기 위한 배열
# adj[from_node] = [weight, to_node] -> heap 구성시 weight을 우선순위로 계산
adj = [[] for _ in range(N+1)]
visited = [False]*(N+1) # MST 에서는 방문체크를 해야함 (1번만 방문)

for i in range(M):
    a, b, c = map(int, input().split())
    adj[a].append([c, b])
    adj[b].append([c, a])

print(prim(1))

File: 그래프/BOJ-1647.swift
// BOJ 1647 도시 분할 계획 - 그래프 MST 

/// 노드 x 의 최상단 부모를 찾는 함수
func getParent(_ parent: inout [Int], _ x: Int) -> Int {
    if parent[x] == x { return x }
    parent[x] = getParent(&parent, parent[x])
    return parent[x]
}

/// a, b 의 parent 를 합치는 함수 - 일반적으로 작은 값으로 설정
func unionParent(_ parent: inout [Int], _ a: Int, _ b: Int) {
    let a = getParent(&parent, a)
    let b = getParent(&parent, b)
    if a < b {
        parent[b] = a
    } else {
        parent[a] = b
    }
}

/// a, b 가 같은 부모를 가지는지 확인하는 함수 - 같은 그룹인지
func findParent(_ parent: inout [Int], _ a: Int, _ b: Int) -> Bool {
    let a = getParent(&parent, a)
    let b = getParent(&parent, b)
    return a == b
}

// Input
let line = readLine()!.split(separator: " ").map{Int(String($0))!}
let n = line[0], m = line[1]

typealias Edge = (a: Int, b: Int, cost: Int)
var edges = [Edge]()
for _ in 1...m {
    let line = readLine()!.split(separator: " ").map{Int(String($0))!}
    edges.append(Edge(a: line[0], b: line[1], cost: line[2]))
}
edges.sort(by: { $0.cost < $1.cost }) // 경로가 여러개일수 있으므로 오름차순 정렬

// 유니온파인드로 경로 연결하면서 비용 계산하기
var parent = [Int]()
for i in 0...n {
    parent.append(i)
}
var connectCost = 0
var maxCost = 0
for (a, b, cost) in edges {
    if findParent(&parent, a, b) == false {
        unionParent(&parent, a, b)
        connectCost += cost
        maxCost = cost
    }
}
// 가장 비용이 큰 경로를 제외하기
print(connectCost - maxCost)